src/Microsoft.ML.Data/Transforms/CopyColumnsTransform.cs

```diff
@@ -3,25 +3,114 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Reflection;
+using System.Collections.Generic;
+using System.Linq;
 using System.Text;
+using Microsoft.ML.Core.Data;
 using Microsoft.ML.Runtime;
 using Microsoft.ML.Runtime.CommandLine;
 using Microsoft.ML.Runtime.Data;
 using Microsoft.ML.Runtime.EntryPoints;
 using Microsoft.ML.Runtime.Internal.Utilities;
 using Microsoft.ML.Runtime.Model;
 
-[assembly: LoadableClass(CopyColumnsTransform.Summary, typeof(CopyColumnsTransform), typeof(CopyColumnsTransform.Arguments), typeof(SignatureDataTransform),
-    CopyColumnsTransform.UserName, "CopyColumns", "CopyColumnsTransform", CopyColumnsTransform.ShortName, DocName = "transform/CopyColumnsTransform.md")]
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(IDataTransform), typeof(CopyColumnsTransformer),
+    typeof(CopyColumnsTransformer.Arguments), typeof(SignatureDataTransform),
+    CopyColumnsTransformer.UserName, "CopyColumns", "CopyColumnsTransform", CopyColumnsTransformer.ShortName,
+    DocName = "transform/CopyColumnsTransformer.md")]
 
-[assembly: LoadableClass(CopyColumnsTransform.Summary, typeof(CopyColumnsTransform), null, typeof(SignatureLoadDataTransform),
-    CopyColumnsTransform.UserName, CopyColumnsTransform.LoaderSignature)]
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(IDataView), typeof(CopyColumnsTransformer), null, typeof(SignatureLoadDataTransform),
+    CopyColumnsTransformer.UserName, CopyColumnsTransformer.LoaderSignature)]
+
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(CopyColumnsTransformer), null, typeof(SignatureLoadModel),
+    CopyColumnsTransformer.UserName, CopyColumnsTransformer.LoaderSignature)]
+
+[assembly: LoadableClass(CopyColumnsTransformer.Summary, typeof(CopyColumnsRowMapper), null, typeof(SignatureLoadRowMapper),
+    CopyColumnsTransformer.UserName, CopyColumnsRowMapper.LoaderSignature)]
 
 namespace Microsoft.ML.Runtime.Data
 {
-    public sealed class CopyColumnsTransform : OneToOneTransformBase
+    public sealed class CopyColumnsEstimator : IEstimator<CopyColumnsTransformer>
+    {
+        private readonly (string Source, string Name)[] _columns;
+        private readonly IHost _host;
+
+        public CopyColumnsEstimator(IHostEnvironment env, string input, string output) :
+            this(env, new (string, string)[1] { (input, output) })
+        {
+        }
+
+        public CopyColumnsEstimator(IHostEnvironment env, params (string Source, string Name)[] columns)
```

Comment:
>Source [](start = 73, length = 6)

From what I've seen in MSDN documentation on `ValueTuple`, while it's a little strange they seem to eschew normal .NET naming conventions on *objects* when used in this context and instead cleave to .NET naming conventions on *parameters*, since despite being a public member name it is being used kinda as a parameter name. So maybe `source` and `name`?

Reply:
Maybe RoboTom should be updated to be `ValueTuple` aware...

---
In reply to: [212752862](https://github.com/dotnet/machinelearning/pull/706#discussion_r212752862) [](ancestors = 212752862)